File: agcwd.py
# Adaptive gamma correction based on the reference.
# Reference:
#   S. Huang, F. Cheng and Y. Chiu, "Efficient Contrast Enhancement Using Adaptive Gamma Correction With
#   Weighting Distribution," in IEEE Transactions on Image Processing, vol. 22, no. 3, pp. 1032-1041,
#   March 2013. doi: 10.1109/TIP.2012.2226047
# Revised from https://github.com/mss3331/AGCWD/blob/master/AGCWD.m

import numpy as np
import cv2


def agcwd(image, w=0.5):
    is_colorful = len(image.shape) >= 3
    img = extract_value_channel(image) if is_colorful else image
    img_pdf = get_pdf(img)
    max_intensity = np.max(img_pdf)
    min_intensity = np.min(img_pdf)
    w_img_pdf = max_intensity * (((img_pdf - min_intensity) / (max_intensity - min_intensity)) ** w)
    w_img_cdf = np.cumsum(w_img_pdf) / np.sum(w_img_pdf)
    l_intensity = np.arange(0, 256)
    l_intensity = np.array([255 * (e / 255) ** (1 - w_img_cdf[e]) for e in l_intensity], dtype=np.uint8)
    enhanced_image = np.copy(img)
    height, width = img.shape
    for i in range(0, height):
        for j in range(0, width):
            intensity = enhanced_image[i, j]
            enhanced_image[i, j] = l_intensity[intensity]
    enhanced_image = set_value_channel(image, enhanced_image) if is_colorful else enhanced_image
    return enhanced_image


def extract_value_channel(color_image):
    color_image = color_image.astype(np.float32) / 255.
    hsv = cv2.cvtColor(color_image, cv2.COLOR_BGR2HSV)
    v = hsv[:, :, 2]
    return np.uint8(v * 255)


def get_pdf(gray_image):
    height, width = gray_image.shape
    pixel_count = height * width
    hist = cv2.calcHist([gray_image], [0], None, [256], [0, 256])
    return hist / pixel_count


def set_value_channel(color_image, value_channel):
    value_channel = value_channel.astype(np.float32) / 255
    color_image = color_image.astype(np.float32) / 255.
    color_image = cv2.cvtColor(color_image, cv2.COLOR_BGR2HSV)
    color_image[:, :, 2] = value_channel
    color_image = np.array(cv2.cvtColor(color_image, cv2.COLOR_HSV2BGR) * 255, dtype=np.uint8)
    return color_image


def main():
    import argparse
    parser = argparse.ArgumentParser()
    parser.add_argument('--image', dest='img_path')
    args = parser.parse_args()
    img_path = args.img_path
    img = cv2.imread(img_path)
    cv2.imshow('base', img)
    enhanced_image = agcwd(img)
    cv2.imwrite('enhanced.jpg', enhanced_image)
    cv2.imshow('enhanced', enhanced_image)
    cv2.waitKey(0)
    cv2.destroyAllWindows()


if __name__ == '__main__':
    main()

File: agcwd.js
/** 
 * adaptive gamma correction based on the reference.
 * Reference:
 *   S. Huang, F. Cheng and Y. Chiu, "Efficient Contrast Enhancement Using Adaptive Gamma Correction With
 *   Weighting Distribution," in IEEE Transactions on Image Processing, vol. 22, no. 3, pp. 1032-1041,
 *   March 2013. doi: 10.1109/TIP.2012.2226047
 * Revised from
 *      https://github.com/mss3331/AGCWD/blob/master/AGCWD.m
 */
const cv = require('opencv4nodejs')
const tf = require('@tensorflow/tfjs')
require('@tensorflow/tfjs-node')

function _extractValueChannel(normalizedColorImage) {
    const hsv = normalizedColorImage.cvtColor(cv.COLOR_BGR2HSV)
    const hsvChannels = hsv.split()
    const valueChannel = hsvChannels[2]
    return valueChannel.mul(255).convertTo(cv.CV_8U)
}

function _setValueChannel(colorImage, valueChannel) {
    const normalizedValueChannel = valueChannel.convertTo(cv.CV_32F).div(255)
    const normalizedColorImage = colorImage.convertTo(cv.CV_32F).div(255)
    const hsv = normalizedColorImage.cvtColor(cv.COLOR_BGR2HSV)
    const [hue, saturation] = hsv.split().slice(0, 2)
    return new cv.Mat([hue, saturation, normalizedValueChannel]).cvtColor(cv.COLOR_HSV2BGR).mul(255).convertTo(cv.CV_8U)
}

function _getPDF(image) {
    const {
        cols,
        rows
    } = image
    const pixelCount = cols * rows
    const hist = cv.calcHist(image, [{
        channel: 0,
        bins: 256,
        ranges: [0, 255]
    }])
    return hist.convertTo(cv.CV_32F).div(pixelCount)
}

function _enhance(image, pdf) {
    const pdfTensor = tf.tensor(pdf.getDataAsArray())
    const cdfTensor = tf.cumsum(pdfTensor).div(tf.sum(pdfTensor))
    const cdfArray = cdfTensor.squeeze().dataSync()
    const intensityArray = cdfArray.map((value, index) => 255 * (index / 255) ** (1 - value))
    const {
        cols,
        rows
    } = image
    console.log(image)
    const enhancedImage = image.copy()
    console.log(enhancedImage)
    for (let i = 0; i < rows; ++i) {
        for (let j = 0; j < cols; ++j) {
            const intensity = enhancedImage.atRaw(i, j)
            enhancedImage.set(i, j, [intensityArray[intensity]])
        }
    }
    return enhancedImage
}

function _transform(image) {
    const imagePDF = _getPDF(image)
    const {
        minVal,
        maxVal
    } = imagePDF.minMaxLoc()
    const {
        cols,
        rows
    } = imagePDF
    const minMat = new cv.Mat(rows, cols, cv.CV_32FC1, minVal)
    // use `sqrt` as pow(element, weight=0.5)
    // you can try to revised the next line 
    const weightedImagePDF = (imagePDF.sub(minMat).div(maxVal - minVal)).sqrt().mul(maxVal)
    return _enhance(image, weightedImagePDF)
}

function transformByAGCWD(image) {
    let img = image.copy()
    const isColor = img.channels >= 3
    if (isColor) {
        const normalizedColorImage = img.convertTo(cv.CV_32F).div(255)
        img = _extractValueChannel(normalizedColorImage)
    }
    let enhancedImage = _transform(img)
    if (isColor) {
        enhancedImage = _setValueChannel(image, enhancedImage)
    }
    return enhancedImage
}

module.exports = {
    transformByAGCWD
}